### PR TITLE
[3D] Fix double map click registration

### PR DESCRIPTION
**Pull request by loicgasser**
I noticed there was a racing condition in the Tooltip directive between cesium initialisation and `isActive` controlling the sent of the identify requests.

This race becomes really obvious when using the non-mimified of version.
See: https://github.com/geoadmin/mf-geoadmin3/pull/2778
But I could observe this issue in prod as well.

To reproduce the error,
1. Open the app with a queryable layer (in dev mode) baseURL + ?dev3d=true&lang=fr&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.swisstopo.pixelkarte-pk25.metadata&layers_opacity=0.95&X=191502.45&Y=769146.75&zoom=8
2. Switch to 3d
3. Click on an object

@fredj Does it seem like an acceptable fix to you?

### CODE REVIEW

**Comment by gjn** [comment]
Why not keep state of registration inside the registerMapEvent function? Simple flag indicating if already registered, flag is reset und dereg call and multiple register calls would simply bail out if flag is already set.

It would be robust against race-conditions and client code (users of registerMapEven and the deref function) don't need to know about and check for noop.

**Comment by oterral** [comment]
Sorry I can't reproduce the bug.

**Comment by loicgasser** [comment]
you should be able to now in debug mode. it happens every time

**Comment by loicgasser** [comment]
@gjn the problem is that deregMapEvents is overwritten.
It goes first through
https://github.com/geoadmin/mf-geoadmin3/blob/master/src/components/tooltip/TooltipDirective.js#L283
because 3d is not active but there is a queryable layer and then
https://github.com/geoadmin/mf-geoadmin3/blob/master/src/components/tooltip/TooltipDirective.js#L270
when finally cesium is loaded.

**Comment by oterral** [comment]
AH yes my breakpoints made it works.

**Comment by loicgasser** [comment]
exactly ;)

File: src/components/tooltip/TooltipDirective.js [hunk 1]
```diff
@@ -262,13 +263,15 @@ goog.require('ga_topic_service');
...
                     if (enabled) {
-                      deregMapEvents();
+                      deregMapEvents = deregMapEvents();
                       deregGlobeEvents = registerGlobeEvents(scope,
                           findFeatures);
                     } else {
                       deregGlobeEvents();
-                      deregMapEvents = registerMapEvents(scope, findFeatures,
-                          gaBrowserSniffer.mobile);
+                      if (deregMapEvents === angular.noop) {
```
Comment (oterral) [inline src/components/tooltip/TooltipDirective.js]: I think we should be put this test at the begining of  `registerMapEvents`  and doiing the same in `registerGlobeEvents`

**Comment by gjn** [comment]
That was not my point. May point was to keep the registration state inside of the registerMapEvents functions. All checks can then be made inside of registerMapEvents function and the function returned by it. That way, _users_ of those functions can call it as many times as they want it, in any order that they want it and we assure that at any time, only one event is registered.

As it's currently done, responsibility to check state of registrations is passed to the users of those functions.

File: src/components/tooltip/TooltipDirective.js (hunk 1 above)
Comment (loicgasser) [inline src/components/tooltip/TooltipDirective.js]: you mean pass the callback in registerMapEvents and if !== to angular.noop return the function passed initially. yeah i think it's nice. I can adapt.

Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: That's basically my point.

Comment (oterral) [inline src/components/tooltip/TooltipDirective.js]: Yes @loicgasser

File: src/components/tooltip/TooltipDirective.js [hunk 2]
```diff
@@ -278,16 +286,17 @@ goog.require('ga_topic_service');
             scope.$watch('isActive', function(active) {
               if (active) {
                 if (is3dActive()) {
-                  deregGlobeEvents = registerGlobeEvents(scope, findFeatures);
+                  deregGlobeEvents = registerGlobeEvents(scope, findFeatures,
+                      deregGlobeEvents);
                 } else {
                   deregMapEvents = registerMapEvents(scope, findFeatures,
-                      gaBrowserSniffer.mobile);
```
Comment (loicgasser) [inline src/components/tooltip/TooltipDirective.js]: third parameter gaBrowserSniffer.mobile was not used

File: src/components/tooltip/TooltipDirective.js [hunk 3]
```diff
@@ -118,7 +118,10 @@ goog.require('ga_topic_service');
                 updateCursorStyle, 10, false, false);
 
         // Register click/touch/mousemove events on map
-        var registerMapEvents = function(scope, onClick) {
+        var registerMapEvents = function(scope, onClick, deregEvent) {
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: Why not keep state in function scope instead of passing it every time?

**Comment by loicgasser** [comment]
wait a second still not good

File: src/components/tooltip/TooltipDirective.js (hunk 3 above)
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: Something like

```
var registerMapEvents = function() {
    var dereg = angular.noop;
    return function(scope, onClick) {
        if (dereg != angular.noop) {
            return dereg;
        }
        .
        .
        .
        dereg = function() {
            deregMapClick();
            ol.Observable.unByKey(deregPointerMove);
            dereg = angular.noop;
        }
        return dereg;
    };
}();
```

That way, you don't need to track results of dereg calls.
Note: code above not tested.

Comment (loicgasser) [inline src/components/tooltip/TooltipDirective.js]: But it's 2 different calls. I tried anyway and we end up with the same issue. Without an instance or variable outside the function, it won't work.
If you call:
https://github.com/geoadmin/mf-geoadmin3/blob/master/src/components/tooltip/TooltipDirective.js#L283
and then
https://github.com/geoadmin/mf-geoadmin3/blob/master/src/components/tooltip/TooltipDirective.js#L270
dereg will always be equal to `angular.noop`.
I understant it's better to be self contained though.

**Comment by loicgasser** [comment]
This seems to work. please test again and let me know

File: src/components/tooltip/TooltipDirective.js (hunk 3 above)
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: Hmm, `dereg = angular.noop` is only set if somebody actually calls the `deregMapEvents` function (and not when calling the `registerMapEvents`.

Anyway, we could use a module-wide variably by moving `var deregMapEvents = angular.noop` out of link function and into module function and then use this variable always. We always set `deregMapEvents` either inside the register functions or inside the deregister function.

```
var registerMapEvents = function(scope, onClick) {
        if (deregMapEvents != angular.noop) {
            return;
        }
        .
        .
        .
        deregMapEvents = function() {
            deregMapClick();
            ol.Observable.unByKey(deregPointerMove);
            deregMapEvents = angular.noop;
        }
        // Nothing to return, as deregMapEvents is outside of this function scope.
    };
```

Working?

File: src/components/tooltip/TooltipDirective.js [hunk 4]
```diff
@@ -118,7 +118,11 @@ goog.require('ga_topic_service');
                 updateCursorStyle, 10, false, false);
 
         // Register click/touch/mousemove events on map
+        var deregMapListener = angular.noop;
         var registerMapEvents = function(scope, onClick) {
+          if (deregMapListener !== angular.noop) {
+            return deregMapListener;
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: No need to return it, as it's upper scope variable already.

File: src/components/tooltip/TooltipDirective.js [hunk 5]
```diff
@@ -144,14 +148,20 @@ goog.require('ga_topic_service');
               updateCursorStyleDebounced(map, evt.pixel);
             });
           }
-          return function() {
+          deregMapListener = function() {
             deregMapClick();
             ol.Observable.unByKey(deregPointerMove);
+            deregMapListener = angular.noop;
           };
+          return deregMapListener;
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: dito.

File: src/components/tooltip/TooltipDirective.js [hunk 6]
```diff
@@ -144,14 +148,20 @@ goog.require('ga_topic_service');
...
         };
 
         // Register leftclick event on globe
+        var deregGlobeListener = angular.noop;
         var registerGlobeEvents = function(scope, onClick) {
+          if (deregGlobeListener != angular.noop) {
+            return deregGlobeListener;
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: dito

File: src/components/tooltip/TooltipDirective.js [hunk 7]
```diff
@@ -168,12 +178,15 @@ goog.require('ga_topic_service');
...
+          deregGlobeListener = function() {
             if (!handler.isDestroyed()) {
-              handler.removeInputAction(Cesium.ScreenSpaceEventType.LEFT_CLICK);
+              handler.removeInputAction(
+                  Cesium.ScreenSpaceEventType.LEFT_CLICK);
               handler.destroy();
+              deregGlobeListener = angular.noop;
             }
           };
+          return deregGlobeListener;
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: dito

File: src/components/tooltip/TooltipDirective.js [hunk 8]
```diff
@@ -267,8 +280,7 @@ goog.require('ga_topic_service');
                           findFeatures);
                     } else {
                       deregGlobeEvents();
-                      deregMapEvents = registerMapEvents(scope, findFeatures,
-                          gaBrowserSniffer.mobile);
+                      deregMapEvents = registerMapEvents(scope, findFeatures);
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: No need to assign here, as it's done inside the registerMapEvents function

File: src/components/tooltip/TooltipDirective.js [hunk 9]
```diff
@@ -280,8 +292,7 @@ goog.require('ga_topic_service');
                 if (is3dActive()) {
                   deregGlobeEvents = registerGlobeEvents(scope, findFeatures);
                 } else {
-                  deregMapEvents = registerMapEvents(scope, findFeatures,
-                      gaBrowserSniffer.mobile);
+                  deregMapEvents = registerMapEvents(scope, findFeatures);
```
Comment (gjn) [inline src/components/tooltip/TooltipDirective.js]: dito

**Comment by loicgasser** [comment]
no you need 2 variables, try https://github.com/geoadmin/mf-geoadmin3/compare/gal_gjin_popup_fix?expand=1 it doesn't work otherwise.

**Comment by loicgasser** [comment]
sorry forget my comment

**Comment by gjn** [comment]
Yes, one for each event (map and globe). Code @ your link looks good to me.

**Comment by loicgasser** [comment]
ok updated I think we got. ;)

**Comment by gjn** [comment]
LGTM. Thanks alot for your patience.

**Pull request merged**